docs-src/concepts.md

```diff
@@ -112,42 +114,123 @@ TODO
 
 ### GatewayClass
 
-`GatewayClass` is cluster-scoped resource defined by the infrastructure ops.
-This resource represents a category of Gateways that can be instantiated.
+`GatewayClass` ([source code][gatewayclass-src]) is cluster-scoped resource
+defined by the infrastructure ops. This resource represents a class of Gateways
+that can be instantiated. 
+
+[gatewayclass-src]: https://github.com/kubernetes-sigs/service-apis/blob/master/api/v1alpha1/gatewayclass_types.go
+
+> Note: this serves the same function as the `networking.IngressClass` resource.
 
 ```yaml
-apiVersion: networking.x-k8s.io/v1alpha1
 kind: GatewayClass
 metadata:
-  name: from-internet
-controller:                          # links to a custom resource
-  apiGroup: networking.acme.io       # that implements this class. 
-  kind: cloud-lb
-  name: from-internet
+  name: cluster-gateway
+spec:
+  controller: "acme.io/gateway-controller"
 ```
 
-A sample controller file:
+We expect that one or more `GatewayClasses` will be created by the infrastructure
+provider for the user. It allows decoupling of which mechanism (e.g. controller)
+implements the `Gateways` from the user. For instance, an infrastructure
+provider may create two `GatewayClasses` named `internet`, `private` to reflect
+`Gateways` that define Internet-facing vs private, internal applications.
 
 ```yaml
-apiVersion: networking.acme.io/v1alpha1
-kind: CloudLB
+kind: GatewayClass
+metadata:
+  name: internet
+  ...
+---
+kind: GatewayClass
 metadata:
-  name: from-internet
-networkTier: auto
-ipAllocation: auto
+  name: private
+  ...
 ```
 
-We expect that one or more GatewayClasses will be created by the infrastructure
-provider for the user as part of their cloud infrastructure. If the cluster
-operator wishes to customize their `GatewayClasses`, they are free to do so,
-interacting with the custom resource parameters. Example parameters
-`GatewayClass.controller` resource will include cluster-level defaults and
-constraints.
+The user of the classes will not need to know *how* `internet` and `private` are
+implemented, whether they are cluster-based or cloud-based LBs.
 
 > The community may converge around a common set of GatewayClass names with
 > well-known meanings (e.g. `internet`, `private`) to ease interoperability
 > between platforms.
 
+#### GatewayClass parameters
+
+Providers of the `Gateway API` may need to pass parameters to their controller
+as part of the class definition. This done using the
+`GatewayClass.spec.parameters` field:
+
+```yaml
+# GatewayClass for Gateways that define Internet-facing applications.
+kind: GatewayClass
+metadata:
+  name: internet
+spec:
+  controller: "acme.io/gateway-controller"
```

Comment:
I wanted to add this comment in the Deployment models section but since that's not part of the diff, I can't do it.

Let me explain what I understand:

I understand that a Gateway Class represents an abstract concept and then the controller defines the vendor for it. And then each Gateway is associated with a `GatewayClass`.

Let's assume that each Gateway corresponds to a proxy container running somewhere, meaning it can be a cloud LB, or it can be a proxy vendor. CMIIW.

First, a GatewayClass is created. Nothing happens here other than validation. No new processes are spawned up or no configuration is changed anywhere. 

The cloud provider case here seems very clear. There will be a controller run by the cloud provider, which spins up a LB whenever the user creates a Gateway with the reference to the specific GatewayClass.

The case of proxy vendor needs more discussion.

From the current doc, it seems like the idea is to first installed a controller that manages the installation and life-cycle of the vendor-specific Ingress Controllers. This controller first installs or configures vendor-specific Ingress Controllers (based on the GatewayClass type).

And the IngressController then takes care of satisfying the rules that are defined by other CRDs.

If this is the case, that's far too complex.

For proxy vendors, I'd propose rather propose the following:
- GatewayClass is created; nothing happens at all (not even validation)
- Vendor-specific Ingress Controller is installed, it reads it's gateway class from config options, and then configures itself accordingly
- The Ingress Controller then binds itself to a single or multiple gateways associated with the GatewayClass.

Reply:
I get where you are coming from, the GatewayClass implementation is very clear from a cloud provider point of view, but is less clear from what are currently Ingress controller's points of view.

My understanding is:

GatewayClass is analogous to the current `ingress-class` annotation to an extent - creating a GatewayClass allows Gateways to say what controller should be responsible for them.

This means that the name of the GatewayClass is actually only important as the thing that gets entered into the Gateway. Having well-defined names allows having sensible defaults for the Gateway to choose from, that's all.

If the infrastructure ops for a cluster wanted to change the default `internet` GatewayClass to be served by a different controller, they can do that by updating `spec.controller`. Or, they can create a new `acme-gateway-controller` GatewayClass to represent the Gateways managed by that controller. I agree that this seems like a lot of complexity.

Regardless, I think both of our concerns would be best addressed by writing up a mapping between the current Ingress concepts and the new ones. I've added an agenda item for today's call (16th Jan 2020).

Reply:
@hbagdi:
- Feel free to open a PR against that section in the deployments to gather your comments
- I think your flow as you have given doesn't conflict w/ what is here; would like to chat about it more

@youngnick 
- Yes, please put on the agenda

Reply:
@hbagdi : I think I understand your question: it is whether or not intended that /GatewayClass itself/ is going to be coupled the lifecycle of a controller, e.g. it necessitates a controller-controller (it's controllers all the way down...)

The answer is no. What I see the typical workflow for proxy vendor would be:

* User installs Gateway CRDs
* User installs ACME.IO proxy software; it comes with a GatewayClass CRD with `{name: acme-gateway, provider: acme.io/gateway}` along w/ default settings. This idea here is that lots of existing ingress providers already have notions about having a configmap to configure their software -- this gives an "official" location for such configuration.
* User creates gateways with class `acme-gateway`

There is no need for a controller that watches gatewayclass and instantiates controllers for each class. It is assumes that you could configure several classes for a given provider if you have different configurations for each.

Reply:
A lot was happening last Friday. Re-reading the comments, these makes sense and is exactly how I wish this turn out. Thanks for the clarifications. Consider this thread as resolved (Github is not showing the Resolve conversation button for some reason).